gan/gan_deepspeed_train.py

```diff
@@ -0,0 +1,177 @@
+            dataset = dset.MNIST(root=args.dataroot, download=True,
+                            transform=transforms.Compose([
+                                transforms.Resize(args.imageSize),
+                                transforms.ToTensor(),
+                                transforms.Normalize((0.5,), (0.5,)),
+                            ]))
+            nc=1
+
+    elif args.dataset == 'fake':
+        dataset = dset.FakeData(image_size=(3, args.imageSize, args.imageSize),
+                                transform=transforms.ToTensor())
+        nc=3
+
+    elif args.dataset == 'celeba':
+        dataset = dset.ImageFolder(root=args.dataroot,
+                           transform=transforms.Compose([
+                               transforms.Resize(args.imageSize),
+                               transforms.CenterCrop(args.imageSize),
+                               transforms.ToTensor(),
+                               transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
+                           ]))
+        nc = 3
+
+    assert dataset
+    return dataset, nc
+
+def train(args):
+    writer = SummaryWriter(log_dir=args.tensorboard_path)
+    create_folder(args.outf)
+    set_seed(args.manualSeed)
+    cudnn.benchmark = True
+    dataset, nc = get_dataset(args)
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=args.batchSize, shuffle=True, num_workers=int(args.workers))
+    torch.cuda.set_device(args.local_rank)
+    device = torch.device("cuda", args.local_rank) #torch.device("cuda:0" if args.cuda else "cpu")
+    ngpu = 0
+    nz = int(args.nz)
+    ngf = int(args.ngf)
+    ndf = int(args.ndf)
+
+    netG = Generator(ngpu, ngf, nc, nz).to(device)
+    netG.apply(weights_init)
+    if args.netG != '':
+        netG.load_state_dict(torch.load(args.netG))
+
+    netD = Discriminator(ngpu, ndf, nc).to(device)
+    netD.apply(weights_init)
+    if args.netD != '':
+        netD.load_state_dict(torch.load(args.netD))
+
+    criterion = nn.BCELoss()
+
+    fixed_noise = torch.randn(args.batchSize, nz, 1, 1, device=device)
+    real_label = 1
+    fake_label = 0
+
+    # setup optimizer
+    optimizerD = torch.optim.Adam(netD.parameters(), lr=args.lr, betas=(args.beta1, 0.999))
+    optimizerG = torch.optim.Adam(netG.parameters(), lr=args.lr, betas=(args.beta1, 0.999))
+
+    model_engineD, optimizerD, _, _ = deepspeed.initialize(args=args, model=netD, model_parameters=netD.parameters(), optimizer=optimizerD)
+    model_engineG, optimizerG, _, _ = deepspeed.initialize(args=args, model=netG, model_parameters=netG.parameters(), optimizer=optimizerG)
+
+    for epoch in range(args.epochs):
+        for i, data in enumerate(dataloader, 0):
+            ############################
+            # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
+            ###########################
+            # train with real
+            netD.zero_grad()
+            real_cpu = data[0].to(device)

```

Comment:
give "real_cpu" another name? it is not a cpu tensor.

Reply:
Sure!